src/python/pants/util/memo.py

```diff
@@ -55,7 +65,9 @@ def per_instance(*args, **kwargs):
   return equal_args(*instance_and_rest, **kwargs)
 
 
-def memoized(func=None, key_factory=equal_args, cache_factory=dict):
+def memoized(
+  func: Optional[F] = None, *, key_factory=equal_args, cache_factory=dict,
```

Comment:
I think this might be a breaking API change? It might be annoying to formulate this as a `deprecated_conditional()` if `key_factory` or `cache_factory` are specified as positional arguments, but we might be able to make some utility for that.

Reply:
Bummer. I think you're right. I removed the kwarg only.

I'm going to pass on a deprecation cycle here because I want to spend the time to instead make more progress on MyPy. Agreed that a utility could be helpful.

Reply:
I'm great with this!